Initiative numbers

lkpttn opened the pull request:
This PR removes the drag and drop list style for a field and number based initiative order. It actually updates the state as a user types and re-renders the list in reverse order when the user clicks Sort.

It load Deacon's Decoys first but technically works for the Spicy Boyz and just assigns them to the Decoy state array. It works fine but not the most clean execution.

![screenshot_2018-10-15 take initiative](https://user-images.githubusercontent.com/1865372/46975861-be86a600-d095-11e8-8117-757314331423.png)

Commit:
Removes the draggable list and maps the array out instead

Next up is assigning initiative numbers and ordering the array based on that.

Commit:
Basic array sorting based on initiative state value

Commit:
New field for initiative

Commit:
Changed to arrow function to allow for `this`

lkpttn committed:
Begin to pass initative up to the parent

Still need to figure out how to target the right player to update their initiative

lkpttn committed:
Successfully passing and changing initiative values in state

Need to rework the onBlur method for triggering updates.

lkpttn committed:
Stopped extraneous updates

Commit:
Merge branch 'initiative-numbers' of https://github.com/SpicyBoyz/grimoire into initiative-numbers

Commit:
Swaps out live sort for true sort button functionality

Also adds dexterity values to characters for placeholder text.

DFurnes commented on src/App.js:
```diff
@@ -109,6 +116,18 @@ class App extends Component {
+    );
+
+    this.setState({
+      playerStateIndex: player,
```
Instead of modifying the `player` argument (which is passed by reference, and so might end up modifying something you're not expecting it to elsewhere), use destructuring here to create a new object with the properties of `player` and the new initiative:

```js
this.setState({
    playerStateIndex: {
        ...this.state[playerStateIndex],
        initiative: initiative,
    },
});
```

lkpttn replied:
You're right, the player reference initiative being changed was actually what was holding together the updating of the fields in List. hmmmmm.

DFurnes commented on src/App.js:
```diff
@@ -109,6 +116,18 @@ class App extends Component {
+  updateInitiative(initiative, player) {
+    console.log('Set ' + player.title + ' initiative to ' + initiative);
+    player.initiative = initiative;
+    const playerStateIndex = this.state.players.find(
```
It could be helpful to send this along when calling this method from `List` so you don't have to iterate over the whole object to find it, e.g. `updateInitiative(initiative, playerId)`.

DFurnes commented on src/components/List/List.js:
```diff
@@ -1,40 +1,52 @@
+    let initiative = event.target.value;
+
+    this.props.handleInitiativeChange(initiative, player);
+    event.preventDefault();
```
I don't think it makes a difference, but the convention I've usually seen is to place this at the top of the method.

DFurnes commented on src/components/List/List.js:
```diff
@@ -1,40 +1,52 @@
+
+  sortPlayers() {
+    // sort by value
+    let initiativeOrderArray = this.props.players.sort(function(a, b) {
```
The `Array.prototype.sort` method is a little tricky since it sorts the array in-place, and so would be trying to modify the props given to this component here! 🙅‍♂️ This might be where you were running into some weird behavior with sorting perhaps... instead, I'd make a copy of the array and sort that:

```js
let initiativeOrderArray = this.props.players.concat().sort(function(a, b) {
  return a.initiative - b.initiative;
});

```

DFurnes replied:
This could also be a nice place to use Lodash's [`sortBy`](https://lodash.com/docs/4.17.10#sortBy):

```js
import { sortBy } from 'lodash';

let initiativeOrderArray = sortBy(this.props.players, 'initiative');
```

DFurnes reviewed (approved):
wheeee 🐉

Commit:
Remove extraneous packages

Commit:
Adds lodash

Commit:
Update updating/sorting logic and display

Moving the sorting method into the parent App. This should clean up some state and prop issues.

Also changed the `updateInitiative` method to not modify the player reference and instead just clone the state, update the proper initiative and then `setState`

Commit:
remove lodash lol

Commit:
Quick padding
